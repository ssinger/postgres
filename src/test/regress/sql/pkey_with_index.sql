
-- Tests for 'constraints with existing index' feature
CREATE TABLE cwi_test( a int , b varchar(10), c char);

-- add some data so that all tests have something to work with.

INSERT INTO cwi_test VALUES( 1, 2 );
INSERT INTO cwi_test VALUES( 3, 4 );
INSERT INTO cwi_test VALUES( 5, 6 );

CREATE UNIQUE INDEX cwi_uniq_idx ON cwi_test(a , b);
ALTER TABLE cwi_test ADD primary key USING INDEX cwi_uniq_idx;

CREATE UNIQUE INDEX cwi_uniq2_idx ON cwi_test(b , a);
ALTER TABLE cwi_test DROP CONSTRAINT cwi_uniq_idx,
	ADD CONSTRAINT cwi_replaced_pkey PRIMARY KEY(b, a)
		WITH (INDEX = 'cwi_uniq2_idx');

DROP INDEX cwi_uniq_idx;
DROP INDEX cwi_uniq2_idx;

CREATE UNIQUE INDEX cwi_uniq_idx ON cwi_test(a , b);
ALTER TABLE cwi_test ADD UNIQUE(a, b) WITH (INDEX = 'cwi_uniq_idx');

ALTER TABLE cwi_test ADD PRIMARY KEY (a, b) WITH ( INDEX = 3 );
ALTER TABLE cwi_test ADD PRIMARY KEY (a, b) WITH ( INDEX = del );
ALTER TABLE cwi_test ADD PRIMARY KEY(b, a)
	WITH (INDEX = 'cwi_idx_doesnt_exist');

ALTER TABLE cwi_test ADD UNIQUE (a);
ALTER TABLE cwi_test ADD PRIMARY KEY(a) WITH (INDEX = 'cwi_test_a_key');

CREATE INDEX cwi_idx1 ON cwi_test(a , b);

-- should fail; non-unique
ALTER TABLE cwi_test ADD primary key(a, b) WITH (INDEX = 'cwi_idx1');

CREATE UNIQUE INDEX cwi_idx2 ON cwi_test(a , b);

-- should fail; WITH INDEX option specified more than once.
ALTER TABLE cwi_test ADD PRIMARY KEY (a, b)
	WITH ( INDEX = 'cwi_idx2', INDEX = 'cwi_idx2' );

ALTER TABLE cwi_test ADD primary key(a, b) WITH (INDEX = 'cwi_idx2');

ALTER TABLE cwi_test DROP CONSTRAINT cwi_test_pkey;

ALTER TABLE cwi_test ADD primary key(a, b);

CREATE UNIQUE INDEX cwi_idx3 ON cwi_test(a);

-- should fail
ALTER TABLE cwi_test DROP CONSTRAINT cwi_test_pkey, ADD PRIMARY KEY(a, b)
	WITH (INDEX = 'cwi_idx3');

DROP INDEX cwi_idx3;

SELECT relname, relkind FROM pg_class WHERE relname like E'rpi\\_%'
	ORDER BY relkind DESC, relname ASC;

DROP TABLE cwi_test;

SELECT relname, relkind FROM pg_class WHERE relname like E'rpi\\_%';

