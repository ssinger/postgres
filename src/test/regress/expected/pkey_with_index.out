-- Tests for 'constraints with existing index' feature
CREATE TABLE cwi_test( a int , b varchar(10), c char);
-- add some data so that all tests have something to work with.
INSERT INTO cwi_test VALUES( 1, 2 );
INSERT INTO cwi_test VALUES( 3, 4 );
INSERT INTO cwi_test VALUES( 5, 6 );
CREATE UNIQUE INDEX cwi_uniq_idx ON cwi_test(a , b);
ALTER TABLE cwi_test ADD primary key(a, b) WITH (INDEX = 'cwi_uniq_idx');
NOTICE:  ALTER TABLE / ADD CONSTRAINT WITH will rename index "cwi_uniq_idx" to "cwi_test_pkey"
CREATE UNIQUE INDEX cwi_uniq2_idx ON cwi_test(b , a);
ALTER TABLE cwi_test DROP CONSTRAINT cwi_test_pkey,
	ADD CONSTRAINT cwi_replaced_pkey PRIMARY KEY(b, a)
		WITH (INDEX = 'cwi_uniq2_idx');
NOTICE:  ALTER TABLE / ADD CONSTRAINT WITH will rename index "cwi_uniq2_idx" to "cwi_replaced_pkey"
DROP INDEX cwi_uniq_idx;
ERROR:  index "cwi_uniq_idx" does not exist
DROP INDEX cwi_uniq2_idx;
ERROR:  index "cwi_uniq2_idx" does not exist
CREATE UNIQUE INDEX cwi_uniq_idx ON cwi_test(a , b);
ALTER TABLE cwi_test ADD UNIQUE(a, b) WITH (INDEX = 'cwi_uniq_idx');
NOTICE:  ALTER TABLE / ADD CONSTRAINT WITH will rename index "cwi_uniq_idx" to "cwi_test_a_b_key"
ALTER TABLE cwi_test ADD PRIMARY KEY (a, b) WITH ( INDEX = 3 );
ERROR:  syntax error
DETAIL:  WITH INDEX option in a PRIMARY KEY/UNIQUE constraint should be a string value.
ALTER TABLE cwi_test ADD PRIMARY KEY (a, b) WITH ( INDEX = del );
ERROR:  syntax error
DETAIL:  WITH INDEX option in a PRIMARY KEY/UNIQUE constraint should be a string value.
ALTER TABLE cwi_test ADD PRIMARY KEY(b, a)
	WITH (INDEX = 'cwi_idx_doesnt_exist');
ERROR:  relation "cwi_idx_doesnt_exist" not found
ALTER TABLE cwi_test ADD UNIQUE (a);
NOTICE:  ALTER TABLE / ADD UNIQUE will create implicit index "cwi_test_a_key" for table "cwi_test"
ALTER TABLE cwi_test ADD PRIMARY KEY(a) WITH (INDEX = 'cwi_test_a_key');
ERROR:  index "cwi_test_a_key" is associated with a constraint
CREATE INDEX cwi_idx1 ON cwi_test(a , b);
-- should fail; non-unique
ALTER TABLE cwi_test ADD primary key(a, b) WITH (INDEX = 'cwi_idx1');
ERROR:  "cwi_idx1" is not a unique index
DETAIL:  Cannot create PRIMARY KEY/UNIQUE constraint using a non-unique index.
CREATE UNIQUE INDEX cwi_idx2 ON cwi_test(a , b);
-- should fail; WITH INDEX option specified more than once.
ALTER TABLE cwi_test ADD PRIMARY KEY (a, b)
	WITH ( INDEX = 'cwi_idx2', INDEX = 'cwi_idx2' );
ERROR:  only one WITH INDEX option can be specified for a PRIMARY KEY/UNIQUE constraint
ALTER TABLE cwi_test ADD primary key(a, b) WITH (INDEX = 'cwi_idx2');
NOTICE:  ALTER TABLE / ADD CONSTRAINT WITH will rename index "cwi_idx2" to "cwi_test_pkey"
ERROR:  multiple primary keys for table "cwi_test" are not allowed
ALTER TABLE cwi_test DROP CONSTRAINT cwi_test_pkey;
ERROR:  constraint "cwi_test_pkey" of relation "cwi_test" does not exist
ALTER TABLE cwi_test ADD primary key(a, b);
ERROR:  multiple primary keys for table "cwi_test" are not allowed
CREATE UNIQUE INDEX cwi_idx3 ON cwi_test(a);
-- should fail
ALTER TABLE cwi_test DROP CONSTRAINT cwi_test_pkey, ADD PRIMARY KEY(a, b)
	WITH (INDEX = 'cwi_idx3');
ERROR:  constraint "cwi_test_pkey" of relation "cwi_test" does not exist
DROP INDEX cwi_idx3;
SELECT relname, relkind FROM pg_class WHERE relname like E'cwi\\_%'
	ORDER BY relkind DESC, relname ASC;
      relname      | relkind 
-------------------+---------
 cwi_test          | r
 cwi_idx1          | i
 cwi_idx2          | i
 cwi_replaced_pkey | i
 cwi_test_a_b_key  | i
 cwi_test_a_key    | i
(6 rows)

DROP TABLE cwi_test;
SELECT relname, relkind FROM pg_class WHERE relname like E'cwi\\_%';
 relname | relkind 
---------+---------
(0 rows)

